File: projetos-fenix-para-estudo/t2ti-erp-fenix-banco-agencia-pessoa-completos/fontes/backend/c-sharp/T2TiERPFenix/T2TiERPFenix/Controllers/Cadastros/PessoaController.cs
using Microsoft.AspNetCore.Mvc;
using System;
using System.Collections.Generic;
using T2TiERPFenix.Models;
using T2TiERPFenix.Repository;

namespace T2TiERPFenix.Controllers
{
    [Route("pessoa")]
    [Produces("application/json")]
    public class PessoaController : Controller
    {
        private IRepositoryWrapper _repository;

        public PessoaController(IRepositoryWrapper repository)
        {
            _repository = repository;
        }

        [HttpGet]
        public IActionResult ConsultarListaPessoa([FromQuery]string filter)
        {
            try
            {
                IEnumerable<Pessoa> lista;
                if (filter == null)
                {
                    lista = _repository.Pessoa.ConsultarLista();
                }
                else
                {
                    // define o filtro
                    Filtro filtro = new Filtro(filter);
                    lista = _repository.Pessoa.ConsultarListaFiltro(filtro);
                }
                return Ok(lista);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Consultar Lista Pessoa]", ex));
            }
        }

        [HttpGet("{id}", Name = "ConsultarObjetoPessoa")]
        public IActionResult ConsultarObjetoPessoa(int id)
        {
            try
            {
                var objeto = _repository.Pessoa.ConsultarObjeto(id);

                if (objeto == null)
                {
                    return StatusCode(404, new RetornoJsonErro(404, "Registro não localizado [Consultar Objeto Pessoa]", null));
                }
                else
                {
                    return Ok(objeto);
                }
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Consultar Objeto Pessoa]", ex));
            }
        }

        [HttpPost]
        public IActionResult InserirPessoa([FromBody]Pessoa objJson)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto inválido [Inserir Pessoa]", null));
                }
                _repository.Pessoa.Inserir(objJson);

                return CreatedAtRoute("ConsultarObjetoPessoa", new { id = objJson.Id }, objJson);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Inserir Pessoa]", ex));
            }
        }

        [HttpPut("{id}")]
        public IActionResult AlterarPessoa([FromBody]Pessoa objJson, int id)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto inválido [Alterar Pessoa]", null));
                }

                if (objJson.Id != id)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto inválido [Alterar Pessoa] - ID do objeto difere do ID da URL.", null));
                }

                var objBanco = _repository.Pessoa.ConsultarObjeto(objJson.Id);

                if (objBanco == null)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto com ID inválido [Alterar Pessoa]", null));
                }

                _repository.Pessoa.Alterar(objBanco, objJson);

                return ConsultarObjetoPessoa(id);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Alterar Pessoa]", ex));
            }
        }

        [HttpDelete("{id}")]
        public IActionResult ExcluirPessoa(int id)
        {
            try
            {
                var objeto = _repository.Pessoa.ConsultarObjeto(id);

                _repository.Pessoa.Excluir(objeto);

                return Ok();
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Excluir Pessoa]", ex));
            }
        }

    }
}

File: projetos-fenix-para-estudo/t2ti-erp-fenix-banco-agencia-pessoa-completos/fontes/backend/delphi/dmvc/Controller/Cadastros/PessoaController.pas
unit PessoaController;

interface

uses mvcframework, mvcframework.Commons,
  System.SysUtils,
  MVCFramework.SystemJSONUtils;

type

  [MVCDoc('CRUD Pessoa')]
  [MVCPath('/pessoa')]
  TPessoaController = class(TMVCController)
  public
    [MVCDoc('Retorna uma lista de objetos')]
    [MVCPath('/($filtro)')]
    [MVCHTTPMethod([httpGET])]
    procedure ConsultarLista(Context: TWebContext);

    [MVCDoc('Retorna um objeto com base no ID')]
    [MVCPath('/($id)')]
    [MVCHTTPMethod([httpGET])]
    procedure ConsultarObjeto(id: Integer);

    [MVCDoc('Insere um novo objeto')]
    [MVCPath]
    [MVCHTTPMethod([httpPOST])]
    procedure Inserir(Context: TWebContext);

    [MVCDoc('Altera um objeto com base no ID')]
    [MVCPath('/($id)')]
    [MVCHTTPMethod([httpPUT])]
    procedure Alterar(id: Integer);

    [MVCDoc('Exclui um objeto com base no ID')]
    [MVCPath('/($id)')]
    [MVCHTTPMethod([httpDelete])]
    procedure Excluir(id: Integer);
  end;

implementation

{ TPessoaController }

uses PessoaService, Pessoa, Commons, Filtro;

procedure TPessoaController.ConsultarLista(Context: TWebContext);
var
  FiltroUrl, FilterWhere: string;
  FiltroObj: TFiltro;
begin
  FiltroUrl := Context.Request.Params['filtro'];
  if FiltroUrl <> '' then
  begin
    ConsultarObjeto(StrToInt(FiltroUrl));
    exit;
  end;

  filterWhere := Context.Request.Params['filter'];
  try
    if FilterWhere = '' then
    begin
      Render<TPessoa>(TPessoaService.ConsultarLista);
    end
    else begin
      // define o objeto filtro
      FiltroObj := TFiltro.Create(FilterWhere);
      Render<TPessoa>(TPessoaService.ConsultarListaFiltroValor(FiltroObj.Campo, FiltroObj.Valor));
    end;
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create
        ('Erro no Servidor [Consultar Lista Pessoa] - Exceção: ' + E.Message,
        E.StackTrace, 0, 500);
    end
    else
      raise;
  end;
end;

procedure TPessoaController.ConsultarObjeto(id: Integer);
var
  Pessoa: TPessoa;
begin
  try
    Pessoa := TPessoaService.ConsultarObjeto(id);

    if Assigned(Pessoa) then
      Render(Pessoa)
    else
      raise EMVCException.Create
        ('Registro não localizado [Consultar Objeto Pessoa]', '', 0, 404);
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create
        ('Erro no Servidor [Consultar Objeto Pessoa] - Exceção: ' + E.Message,
        E.StackTrace, 0, 500);
    end
    else
      raise;
  end;
end;

procedure TPessoaController.Inserir(Context: TWebContext);
var
  Pessoa: TPessoa;
begin
  try
    Pessoa := Context.Request.BodyAs<TPessoa>;
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create('Objeto inválido [Inserir Pessoa] - Exceção: ' +
        E.Message, E.StackTrace, 0, 400);
    end
    else
      raise;
  end;

  try
    TPessoaService.Inserir(Pessoa);
    Render(TPessoaService.ConsultarObjeto(Pessoa.Id));
  finally
  end;
end;

procedure TPessoaController.Alterar(id: Integer);
var
  Pessoa, PessoaDB: TPessoa;
begin
  try
    Pessoa := Context.Request.BodyAs<TPessoa>;
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create('Objeto inválido [Alterar Pessoa] - Exceção: ' +
        E.Message, E.StackTrace, 0, 400);
    end
    else
      raise;
  end;

  if Pessoa.Id <> id then
    raise EMVCException.Create('Objeto inválido [Alterar Pessoa] - ID do objeto difere do ID da URL.',
      '', 0, 400);

  PessoaDB := TPessoaService.ConsultarObjeto(Pessoa.id);

  if not Assigned(PessoaDB) then
    raise EMVCException.Create('Objeto com ID inválido [Alterar Pessoa]',
      '', 0, 400);

  try
    if TPessoaService.Alterar(Pessoa) > 0 then
      Render(TPessoaService.ConsultarObjeto(Pessoa.Id))
    else
      raise EMVCException.Create('Nenhum registro foi alterado [Alterar Pessoa]',
        '', 0, 500);
  finally
    FreeAndNil(PessoaDB);
  end;
end;

procedure TPessoaController.Excluir(id: Integer);
var
  Pessoa: TPessoa;
begin
  Pessoa := TPessoaService.ConsultarObjeto(id);

  if not Assigned(Pessoa) then
    raise EMVCException.Create('Objeto com ID inválido [Excluir Pessoa]',
      '', 0, 400);

  try
    if TPessoaService.Excluir(Pessoa) > 0 then
      Render(200, 'Objeto excluído com sucesso.')
    else
      raise EMVCException.Create('Nenhum registro foi excluído [Excluir Pessoa]',
        '', 0, 500);
  finally
    FreeAndNil(Pessoa)
  end;
end;

end.
